agent/structs/config_entry_discoverychain.go

```diff
@@ -0,0 +1,746 @@
+package structs
+
+import (
+	"fmt"
+	"math"
+	"sort"
+	"time"
+
+	"github.com/hashicorp/consul/acl"
+)
+
+// ServiceRouterConfigEntry defines L7 (e.g. http) routing rules for a named
+// service exposed in Connect.
+//
+// This config entry represents the topmost part of the discovery chain. Only
+// one router config will be used per resolved discovery chain and is not
+// otherwise discovered recursively (unlike splitter and resolver config
+// entries).
+//
+// Router config entries will be restricted to only services that define their
+// protocol as http-based (in centralized configuration).
+type ServiceRouterConfigEntry struct {
+	Kind string
+	Name string
+
+	// Routes is the list of routes to consider when processing L7 requests.
+	// The first rule to match in the list is terminal and stops further
+	// evaluation.
+	//
+	// Traffic that fails to match any of the provided routes will be routed to
+	// the default service.
+	Routes []ServiceRoute
+
+	RaftIndex
+}
+
+func (e *ServiceRouterConfigEntry) GetKind() string {
+	return ServiceRouter
+}
+
+func (e *ServiceRouterConfigEntry) GetName() string {
+	if e == nil {
+		return ""
+	}
+
+	return e.Name
+}
+
+func (e *ServiceRouterConfigEntry) Normalize() error {
+	if e == nil {
+		return fmt.Errorf("config entry is nil")
+	}
+
+	e.Kind = ServiceRouter
+
+	// TODO(rb): anything to normalize?
+
+	return nil
+}
+
+func (e *ServiceRouterConfigEntry) Validate() error {
+	if e.Name == "" {
+		return fmt.Errorf("Name is required")
+	}
+
+	// TODO(rb): enforce corresponding service has protocol=http
```

Comment:
I think this one is the most important for MVP as it will certainly catch people out if they set all this up against a non-http service and nothing works. It should be easy enough in the basic case though of looking up the service-defaults for the same target. I think that's enough for now without resolving whole possible chain.

It would also be easy enough to check the protocol on all mentioned services in this policy though. I don't know if we need to walk the graph further since the splitter policies lower down should also do same validation anyway?

Reply:
Note that we also have to do the reverse check: if you edit the default protocol for a service we have to verify it doesn't break any existing discovery chain.